src/display.rs

```diff
@@ -78,6 +78,65 @@ impl Display for InlineTable {
 }
 
 impl Table {
+    fn visit_nested_tables_in_original_order<'t, F>(
+        &'t self,
+        path: &mut Vec<&'t str>,
+        is_array_of_tables: bool,
+        callback: &mut F,
+    ) -> Result
+    where
+        F: FnMut(&Table, &Vec<&'t str>, bool) -> Result,
+    {
+        let mut positions = vec![];
+        self.visit_nested_tables(path, is_array_of_tables, &mut |t, _, _| {
+            if let Some(pos) = t.position {
+                positions.push(pos);
+            }
+            Ok(())
+        })?;
+        positions.sort();
+        let mut position_iter = positions.iter();
+
+        // If a table has a .position set then we calculate whether we should
+        // print based on whether it matches current_position. If .position is
+        // None then it was added programatically and we decide whether to print
+        // it based on whether we printed the previous table we visited. This
+        // is to avoid printing tables more than once.
+        //
+        // We set should_print to true the first time we went around the
+        // loop, so initially None-positioned Tables will have already been
+        // printed.
+        let mut should_print = true;
+
+        let mut current_position: Option<&usize> = position_iter.next();
+        while current_position.is_some() {
```

Comment:
The complexity of this algorithm is quadratic in the worst case. I think we can do better here. We could serialize the document to `Vec<(u64, String)>` in one pass of `visit_nested_tables`. Each entry of this vec would represent a table. The `u64` will be an id of the table sorted as in the final `String`, and the second part of the tuple is the `String` representation of the table. Then all we need to do is to sort the vec by id and concatenate the strings.

How do we construct this vec? Something like this:
```rust
       // this number should be at least the number of tables with position `None`
       // alternatively, this could be computed in one pass of `visit_nested_tables`
       const MULTIPLIER: usize = 100500;
       // this would account for all `None` positions
       let mut increment = 0;
       let mut last_position = 0;
       let mut tables = Vec::new();
       self.visit_nested_tables(path, is_array_of_tables, &mut |t, p, is_array| {
            if let Some(pos) = t.position {
                last_position = pos;
                increment = 0;
            } else {
                increment += 1;
            }
            let id = last_position * MULTIPLIER + increment;
            let mut s = String::new();
            visit_table(&mut s, t, p, is_array);
            tables.push((id, s));
            Ok(())
        })?;
        tables.sort_by_key(|&(id, _)| id);
```
This wouldn't work for a generic callback though, so it would require changing the type signature.

Reply:
Even better, since `sort` is stable sort, we can simplify the algorithm:
```rust
       let mut last_position = 0;
       let mut tables = Vec::new();
       self.visit_nested_tables(path, is_array_of_tables, &mut |t, p, is_array| {
            if let Some(pos) = t.position {
                last_position = pos;
            }
            let mut s = String::new();
            visit_table(&mut s, t, p, is_array);
            tables.push((last_position, s));
            Ok(())
        })?;
        tables.sort_by_key(|&(id, _)| id);
```

Reply:
Aha! For some reason I didn't think about sorting tuples of position and payload (even though that's exactly what I would do in python). Thanks for the tip.

I'm wondering if there is a payload that we could use that avoids the extra string copy in the happy case, but a Table can't serialise itself without knowing it's path, so String is probably the best payload.

I'm away from the computer this weekend, but I will try to code this up during the week.